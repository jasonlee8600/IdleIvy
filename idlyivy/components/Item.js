import Image from "next/image";
import { useWeb3React } from "@web3-react/core";
import { getContract } from "../conect/yaleContract";
import { useState, useEffect } from "react";
import Pending from '@/components/Pending'

function Item({ title, desc, image, busi, busiNum, init, balance, user}) {
    
  //Consts for user wallet connection and business info
  const web3reactContext = useWeb3React();
  const [cost, setCost] = useState(2)
  const [busiCost, setBusiCost] = useState(2)
  const [pending, setPending] = useState(false);
  
  useEffect(() => {
    loadBusi()
  });

  //Function to set business information for components
  async function loadBusi() {
    
    if(web3reactContext.account != undefined && user['resets'] > 0) {

      const YaleContract = getContract(
        web3reactContext.library,
        web3reactContext.account
      );
    
     
      if (busi['time'] != 0) {
        var tmpCost = parseInt(await YaleContract.upgradeMultCost(busiNum))
        setCost(tmpCost)
      }
      var nextBusi = parseInt(await YaleContract.nextBusiness())
      if (nextBusi == busiNum) {
        var tmpCost = parseInt(await YaleContract.nextBusinessCost())
        setBusiCost(tmpCost)
      }
      
    }
  }
  
  //Function to upgrade the business
  async function upgradeMult() {
        
    if(web3reactContext.account != undefined && busi['time'] != 0) {
      const YaleContract = getContract(
          web3reactContext.library,
          web3reactContext.account
      );

      var tmpCost = parseInt(await YaleContract.upgradeMultCost(busiNum))
      var mintable = parseInt(await YaleContract.mintableCoin())
        
      if (mintable + balance >= tmpCost) {
        try {
          setPending(true)
          let upgrade = await YaleContract.upgradeMult(busiNum);

          await upgrade.wait();

          init();
          setPending(false)

        }
        catch (error) {
          console.log(error)
          setPending(false)

        }
      }
      else {
        console.log("Not enough coin")
      }
    }
    else {
      console.log("Connect wallet or buy business")
    }
}

//function to unlock next business
async function unlockBusi() {
  
  if(web3reactContext.account != undefined && user['resets'] != 0) {
    const YaleContract = getContract(
        web3reactContext.library,
        web3reactContext.account
    );

    var nextBusi = parseInt(await YaleContract.nextBusiness())
    var tmpCost = parseInt(await YaleContract.nextBusinessCost())
    var mintable = parseInt(await YaleContract.mintableCoin())
      
    if ((nextBusi == busiNum) && (mintable + balance >= tmpCost)) {
      try {
        setPending(true)

        let unlock = await YaleContract.unlockNextBusiness();

        await unlock.wait();

        init();
        setPending(false)

      }
      catch (error) {
        console.log(error)
        setPending(false)

      }
    }
    else {
      console.log("Not enough coin or unlock previous businesses first")
    }
  }
  else {
    console.log("Connect wallet or join game")
  }
}
  
  return (
          <>
            <Pending pending={pending}></Pending>
            <div className='flex flex-row ml-[100px]'>
              <Image alt="Buttery" src={image} width={150} height={90} className="rounded-full outline outline-white outline-[4px]"></Image>
              
              <div className='flex flex-row w-full pl-12'>
                
                <div className='flex flex-row z-0 bg-gradient-to-b from-[#272727] to-[#797979] h-[90px] w-5/6 rounded-full outline outline-white outline-[4px]'>
                  
                    {busi['time'] != 0 ?
                      <button type="button" className="self-start z-10 bg-[#202B64] h-[90px] w-1/4 rounded-full outline outline-white outline-[4px] text-2xl text-white"
                        onClick={upgradeMult}>
                          Buy
                        <h3 className="text-xl">Cost: {cost/10}</h3> 
                      </button>
                    :
                      <button type="button" className="self-start z-10 bg-[#202B64] h-[90px] w-1/4 rounded-full outline outline-white outline-[4px] text-2xl text-white"
                        onClick={unlockBusi}>
                          {busiCost != 2 ?
                            <div>
                              <h3>Buy</h3>
                              <h3 className="text-xl">Cost: {busiCost/10}</h3>
                            </div>
                          : 
                            <h3>Locked</h3>
                          }
                      </button>
                    }
                  
                    <div className='flex flex-col justify-center w-4/5'>
                        <h3 className={`text-2xl text-white text-center`}>{title}: {busi['mult']}</h3>
                        <h4 className={`text-lg text-white text-center`}>{desc}</h4>
                        <h4 className={`text-lg text-white text-center`}>Rate: {busi['mult'] * busi['quantity']} / minute</h4>
                    </div>
                </div>
              </div>
            </div>
          </>

            );
        }

export default Item;