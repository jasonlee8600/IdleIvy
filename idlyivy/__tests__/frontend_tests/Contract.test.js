import React from 'react';
import Contract from '../../pages/contract'
import Item from '../../components/Item'
import {render, screen, cleanup} from '@testing-library/react'
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom'


afterEach(() => {
    cleanup(); 
})

describe("contract.js testing", () => {
    test('renders correctly', () => {
        const tree = renderer.create(<Contract />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
