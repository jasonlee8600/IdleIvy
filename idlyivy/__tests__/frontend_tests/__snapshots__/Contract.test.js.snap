// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract.js testing renders correctly 1`] = `
[
  <div
    className="flex flex-row md:justify-end justify-between sticky top-0  min-h-[75px] w-full z-20 md:border-l-[0px]"
    data-testid="topnav"
  >
    <div
      className="md:hidden flex flex-col pl-4 py-4 justify-between"
    >
      <a
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        <h2
          className="text-white font-semibold text-2xl"
          data-testid="IdleIvy"
        >
          IdleIvy
        </h2>
      </a>
      <a
        href="/leaderboard"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        <h2
          className="text-white font-semibold text-2xl"
          data-testid="Leaderboard"
        >
          Leaderboard
        </h2>
      </a>
    </div>
    <div
      className="px-4 self-center py-1"
    >
      <div
        className="flex flex-col md:flex-row justify-between gap-[20px]"
      >
        <div
          className="relative inline-block text-left w-[200px] h-[40px]"
        >
          <div>
            <button
              className="btn-nav w-full text-xl hover:border-2 dark:border-white border-black"
              onClick={[Function]}
              type="button"
            >
              Connect
              <div
                className="ml-6"
              >
                <img
                  className="w-6 h-6 mt-1"
                  src="drop.png"
                />
              </div>
            </button>
          </div>
        </div>
        <div
          className="btn-nav w-[200px] bg-[#607BEE] hover:bg-[#607BEE] font-semibold"
        >
          <h2>
            Not Connected
          </h2>
        </div>
      </div>
    </div>
  </div>,
  <a>
    Something went wrong
  </a>,
]
`;
